File: Rust/src/lexer.rs
use crate::tokens::Token;

// Used to determine whether a character is a word character.
const ALPHA: &str = "AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz'";

// Scans an input string slice, creating tokens and placing them into a result
// vector.
pub fn lex<'a, 'b>(input: &'a str, result: &'b mut Vec<Token<'a>>) {
    if input.len() == 0 {
        panic!("Input should contain at least one character.");
    }

    if !input.is_ascii() {
        panic!("Input should contain only ASCII characters.");
    }

    let mut lexer: Lexer<'a, 'b> = Lexer::new(input, result);
    lexer.invoke();
}

// Manages state and performs lexigraphical analysis for the lex function.
struct Lexer<'a, 'b> {
    // The string slice that will be tokenized.
    input: &'a str,

    // The vector into which tokens will be placed.
    result: &'b mut Vec<Token<'a>>,

    // A buffer used to track slices of word characters.
    buffer: Option<(usize, usize)>
}

impl<'a, 'b> Lexer<'a, 'b> {
    // Instantiates a new lexer.
    fn new(input: &'a str, result: &'b mut Vec<Token<'a>>) -> Lexer<'a, 'b> {
        Lexer {
            input: input,
            result: result,
            buffer: None
        }
    }

    // Invokes the lexer.
    fn invoke(&mut self) {
        self.result.clear();

        for (i, c) in self.input.chars().enumerate() {
            if ALPHA.contains(c) {
                self.push_alpha(i);
            } else {
                self.push_special(i);
            }
        }

        // If the input string ends in one or more word characters, the buffer
        // will still be referencing some data.
        self.push_buffer();
    }

    // Pushes the index of an alpha character onto the buffer.
    fn push_alpha(&mut self, index: usize) {
        if let Some((min, _)) = self.buffer {
            self.buffer = Some((min, index + 1));
        } else {
            self.buffer = Some((index, index + 1));
        }
    }

    // Converts the index of a special character to a token and pushes it onto
    // the result vector.
    fn push_special(&mut self, index: usize) {
        // A special character is about to be added. If there is a word in the
        // buffer, push it first.
        self.push_buffer();

        let token: Token<'a> = Token::Special(&self.input[index..index + 1]);
        self.result.push(token);
    }

    // Pushes the contents of the buffer, if present, to the result vector and
    // resets the buffer.
    fn push_buffer(&mut self) {
        let Option::Some((start, end)) = self.buffer else {
            // The buffer is empty, so exit.
            return;
        };

        let token: Token<'a> = Token::Word(&self.input[start..end]);
        self.result.push(token);

        self.buffer = None;
    }
}

File: Haskell/app/Lexer.hs
module Lexer (lex') where

import Tokens ( Token(..) )

lex' :: String -> [Token]
lex' text =
  let state0 = State [] []
      state1 = finalize $ foldl push state0 text
   in getTokens state1

push :: State -> Char -> State
push (State [] tokens) c
  | isAlpha c = State [c] tokens
  | otherwise = State [] (Special c : tokens)
push (State buffer tokens) c
  | isAlpha c = State (c : buffer) tokens
  | otherwise =
      let word = reverse buffer
       in State [] (Special c : Word word : tokens)

finalize :: State -> State
finalize s@(State buffer tokens)
  | null buffer = s
  | otherwise =
      let word = reverse buffer
       in State [] (Word word : tokens)

getTokens :: State -> [Token]
getTokens (State _ tokens) = tokens

alpha :: String
alpha = "AaBbCcDdEeFfGgHhIiJjKkLlMmNnOoPpQqRrSsTtUuVvWwXxYyZz'"

isAlpha :: Char -> Bool
isAlpha c = c `elem` alpha

data State = State String [Token]
    deriving(Show)
